src/Swarm/Language/LSP/Hover.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Swarm.Language.LSP.Hover where
+
+import Data.Rope.UTF16 qualified as R
+import Data.Text (Text)
+import Data.Text qualified as T
+import Language.LSP.Types qualified as J
+import Language.LSP.VFS
+import Swarm.Language.Parse (readTerm')
+import Swarm.Language.Syntax
+
+withinBound :: Int -> SrcLoc -> Bool
+withinBound pos (SrcLoc start end) = pos >= start && pos < end
+withinBound _ NoLoc = False
+
+showHoverInfo ::
+  J.NormalizedUri ->
+  J.TextDocumentVersion ->
+  J.Position ->
+  VirtualFile ->
+  Maybe Text
+showHoverInfo _ _ (J.Position myLine myCol) vf@(VirtualFile _ _ myRope) =
+  astSize
+ where
+  content = virtualFileText vf
+  absolutePos = R.rowColumnCodeUnits (R.RowColumn (fromIntegral myLine) (fromIntegral myCol)) myRope
+  astSize = case readTerm' content of
+    Right Nothing -> Nothing
+    Right (Just (Syntax _ term)) -> Just $ explain $ digAt term absolutePos
+    Left _ -> Nothing
+
+descend ::
+  -- | default
+  Term ->
+  -- | position
+  Int ->
+  -- | next element to inspect
+  Syntax ->
+  Term
+descend t pos (Syntax l1 t1) =
+  if withinBound pos l1
+    then digAt t1 pos
+    else t
+
+descend2 ::
+  -- | default
+  Term ->
+  -- | position
+  Int ->
+  -- | next element to inspect
+  Syntax ->
+  -- | alternate element to inspect
+  Syntax ->
+  Term
+descend2 t pos (Syntax l1 t1) s2 =
+  if withinBound pos l1
+    then digAt t1 pos
+    else descend t pos s2
+
+digAt :: Term -> Int -> Term
+digAt t pos = case t of
+  SLam _ _ s -> descend t pos s
+  SApp s1 s2 -> descend2 t pos s1 s2
+  SLet _ _ _ s1 s2 -> descend2 t pos s1 s2
+  SPair s1 s2 -> descend2 t pos s1 s2
+  SDef _ _ _ s -> descend t pos s
+  SBind _ s1 s2 -> descend2 t pos s1 s2
+  SDelay _ s -> descend t pos s
+  x -> x
+
+explain :: Term -> Text
+explain = \case
+  TUnit -> "The unit value."
+  TConst c -> briefDoc $ constDoc $ constInfo c
```

Comment:
Here I utilize the programmatic documentation of `TConst`.  For all of the other terms, I copied and pasted the Haddocks.